engines/wage/combat.cpp

```diff
@@ -0,0 +1,915 @@
+}
+
+void WageEngine::encounter(Chr *player, Chr *chr) {
+	char buf[512];
+
+	snprintf(buf, 512, "You encounter %s%s.", chr->_nameProperNoun ? "" : getIndefiniteArticle(chr->_name),
+				chr->_name.c_str());
+	appendText(buf);
+
+	if (!chr->_initialComment.empty())
+		appendText(chr->_initialComment.c_str());
+
+	if (chr->_armor[Chr::HEAD_ARMOR] != NULL) {
+		snprintf(buf, 512, "%s%s is wearing %s.", chr->getDefiniteArticle(true), chr->_name.c_str(),
+					getIndefiniteArticle(chr->_armor[Chr::HEAD_ARMOR]->_name));
+		appendText(buf);
+	}
+	if (chr->_armor[Chr::BODY_ARMOR] != NULL) {
+		snprintf(buf, 512, "%s is protected by %s%s.", getGenderSpecificPronoun(chr->_gender, true),
+					prependGenderSpecificPronoun(chr->_gender), chr->_armor[Chr::BODY_ARMOR]->_name.c_str());
+		appendText(buf);
+	}
+	if (chr->_armor[Chr::SHIELD_ARMOR] != NULL) {
+		Obj *obj = chr->_armor[Chr::SHIELD_ARMOR];
+
+		snprintf(buf, 512, "%s carries %s%s.", getGenderSpecificPronoun(chr->_gender, true),
+				obj->_namePlural ? "" : getIndefiniteArticle(obj->_name), obj->_name.c_str());
+		appendText(buf);
+	}
+}
+
+void WageEngine::performCombatAction(Chr *npc, Chr *player) {
+	if (npc->_context._frozen)
+		return;
+
+	RandomHat hat(_rnd);
+
+	bool winning = (npc->_context._statVariables[PHYS_HIT_CUR] > player->_context._statVariables[PHYS_HIT_CUR]);
+	int validMoves = getValidMoveDirections(npc);
+	ObjArray *weapons = npc->getWeapons(false);
+	ObjArray *magics = npc->getMagicalObjects();
+	// TODO: Figure out under what circumstances we need to add +1
+	// for the chance (e.g. only when all values were set to 0?).
+	if (winning) {
+		if (!_world->_weaponMenuDisabled) {
+			if (weapons->size() > 0)
+				hat.addTokens(kTokWeapons, npc->_winningWeapons + 1);
+			if (magics->size() > 0)
+				hat.addTokens(kTokMagic, npc->_winningMagic);
+		}
+		if (validMoves != 0)
+			hat.addTokens(kTokRun, npc->_winningRun + 1);
+		if (npc->_inventory.size())
+			hat.addTokens(kTokOffer, npc->_winningOffer + 1);
+	} else {
+		if (!_world->_weaponMenuDisabled) {
+			if (weapons->size() > 0)
+				hat.addTokens(kTokWeapons, npc->_losingWeapons + 1);
+			if (magics->size() > 0)
+				hat.addTokens(kTokMagic, npc->_losingMagic);
+		}
+		if (validMoves != 0)
+			hat.addTokens(kTokRun, npc->_losingRun + 1);
+		if (npc->_inventory.size())
+			hat.addTokens(kTokOffer, npc->_losingOffer + 1);
+	}
+
+	ObjList *objs = &npc->_currentScene->_objs;
+	if (npc->_inventory.size() < npc->_maximumCarriedObjects) {
+		int cnt = 0;
+		for (ObjList::const_iterator it = objs->begin(); it != objs->end(); ++it, ++cnt) {
+			if ((*it)->_type != Obj::IMMOBILE_OBJECT) {
+				// TODO: I'm not sure what the chance should be here.
+				hat.addTokens(cnt, 123);
+			}
+		}
+	}
+
+	int token = hat.drawToken();
+	switch (token) {
+		case kTokWeapons:
+			// TODO: I think the monster should choose the "best" weapon.
+			performAttack(npc, player, weapons->operator[](_rnd->getRandomNumber(weapons->size() - 1)));
```

Comment:
Again, does this actually happen? Is there a test case? (same with kTokMagic below)